controllers/virtualmachine/volumebatch/volumebatch_controller.go

```diff
@@ -567,6 +579,76 @@ func (r *Reconciler) updateVolumeStatusWithPVCInfo(
 	return nil
 }
 
+// isVolumeTrackedByLegacyAttachment checks if a volume is already
+// managed by a legacy CnsNodeVmAttachment. Returns true if the volume
+// is tracked by legacy attachment, false if this volume needs to be
+// added to an existing (or a new) CNSNodeVMBatchAttachment spec.
+func (r *Reconciler) isVolumeTrackedByLegacyAttachment(
+	ctx *pkgctx.VolumeContext,
+	volume vmopv1.VirtualMachineVolume) (bool, error) {
+
+	// Skip non PVC backed volumes.
+	if volume.PersistentVolumeClaim == nil {
+		return false, nil
+	}
+
+	legacyAttachmentName := pkgutil.CNSAttachmentNameForVolume(ctx.VM.Name, volume.Name)
+
+	legacyAttachment := &cnsv1alpha1.CnsNodeVmAttachment{}
+	err := r.Client.Get(ctx, client.ObjectKey{
+		Name:      legacyAttachmentName,
+		Namespace: ctx.VM.Namespace,
+	}, legacyAttachment)
+
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			// No legacy attachment found, this is a greenfield volume attachment.
+			return false, nil
+		}
+
+		// This can fail due to a number of reasons, but we still err
+		// on the side of caution since the flip side is we have
+		// attachments being tracked in two places.
+		return false, fmt.Errorf("failed to check if volume is handled by CnsNodeVmAttachment %s: %w",
+			legacyAttachmentName, err)
+	}
+
+	// Legacy attachment exists, verify it's for the same VM and PVC
+	//
+	// TODO: AKP: Since we are only going to run the new volume
+	// controller, we need to duplicate the logic from
+	// createCNSAttachmentButAlreadyExists in volume controller v1 to
```

Comment:
Not only from `createCNSAttachmentButAlreadyExists `, also `attachmentsToDelete` to find out what volumes to be deleted by comparing volume.spec and nodeAttachment then clean them.